### ISSUE TITLE
Big endian not enforced in float arrays in de/serialisation

### ISSUE THREAD

**Issue by lilyminium** [openforcefield/openff-toolkit#1154]
**Describe the bug**

**Serialisation**
https://github.com/openforcefield/openff-toolkit/blob/ce84d821cd2097dc249abac43de2e61249bcf5d5/openff/toolkit/utils/utils.py#L664-L666

`newbyteorder` does not change the order of the bytes, but only the order that they are interpreted in. e.g.

```python
>>> arr = np.arange(3).astype(float)
>>> arr
array([0., 1., 2.])
>>> arr.newbyteorder(">")
array([0.00000e+000, 3.03865e-319, 3.16202e-322])
```

So this function writes exactly the same bytes in the same order that were in the array before.

**Deserialisation**

https://github.com/openforcefield/openff-toolkit/blob/ce84d821cd2097dc249abac43de2e61249bcf5d5/openff/toolkit/utils/utils.py#L692-L694

`newbyteorder` is not side-effecting, so this doesn't do anything. See https://github.com/openforcefield/openff-interchange/issues/345 for more.

**To Reproduce**

**Output**

**Computing environment (please complete the following information):**
 - Operating system
 - Output of running `conda list` 

**Additional context**

The reason that the endian enforcement has not shown up in tests is because both functions have no effect, so they default to the native endiannness of the system. This likely hasn't been a problem for users transferring files because by and large people work on x86, which is little-endian. I think one could test the functions individually by creating arrays with opposite endians and checking that they are serialized and deserialized properly. e.g. on my Mac

```
>>> import numpy as np
>>> from openff.toolkit.utils.utils import serialize_numpy, deserialize_numpy
>>> import sys
>>> sys.byteorder
'little'
>>> dt_bigendian = np.dtype(float).newbyteorder(">")
>>> arr = np.arange(3).astype(dt_bigendian)
>>> arr
array([0., 1., 2.])
>>> np.frombuffer(arr.tobytes())
array([0.00000e+000, 3.03865e-319, 3.16202e-322])
>>> actually_serialize_numpy = lambda x: (x.astype(dt_bigendian).tobytes(), x.shape)
>>> actually_deserialize_numpy = lambda x, y: np.reshape(np.frombuffer(x, dtype=dt_bigendian), y)
>>> np.frombuffer(actually_serialize_numpy(arr)[0])
array([0.00000e+000, 3.03865e-319, 3.16202e-322])
>>> little_arr = np.arange(3).astype(float)
>>> np.frombuffer(serialize_numpy(little_arr)[0], dtype=dt_bigendian)
array([0.00000e+000, 3.03865e-319, 3.16202e-322])
>>> actually_deserialize_numpy(*actually_serialize_numpy(arr))
array([0., 1., 2.])
>>> actually_deserialize_numpy(np.arange(3).astype(float).tobytes(), (3,))
array([0.00000e+000, 3.03865e-319, 3.16202e-322])
``` 

So tests could look like:

```
def test_serialize_numpy():
	original = np.arange(3).astype(float)
	dt_little = np.dtype(float).newbyteorder("<")
	dt_big = np.dtype(float).newbyteorder(">")
	arr = original.astype(dt_little)
	assert_allclose(arr, original)
	deserialized = np.from_buffer(serialize_numpy(arr)[0], dtype=dt_big)
	assert_allclose(deserialized, original)

def test_deserialize_numpy():
	original = np.arange(3).astype(float)
	dt_big = np.dtype(float).newbyteorder(">")
	arr = original.astype(dt_big)
	deserialized = deserialize_numpy(arr.tobytes(), arr.shape)
	assert_allclose(deserialized, original)

def test_serialization_roundtrip():
	original = np.arange(3).astype(float)
	deserialized = deserialize_numpy(*serialize_numpy(original))
	assert_allclose(deserialized, original)
```

These should fail appropriately on a little-endian system; on a big-endian system, code where native endianness is used instead it might pass / fail silently. I *think* `test_serialize_numpy` should guard against that.

**Comment by lilyminium** [comment]
@mattwthompson I thought a bit about how to test things -- roundtrip tests will only catch if *one* of the de/serialization functions isn't working, and only on a non-native endian architecture, but you can test them separately.

**Comment by mattwthompson** [comment]
This wasn't clicking for me, so I had to pretty much run through everything you did - below are my notes, which should be equivalent in information content to everything you already posted. If I'm understanding everything right, the changes to the source code should be small, but we'd want to add something like those tests to really ensure they it doesn't continue to keep not actually testing endian-ness.

So calling `.newbyteorder` on an array doesn't produce the effect we wanted (make sure the bytes are in the expected order). But calling it on a type does (?)

```python3
In [21]:  arr.newbyteorder("<").tobytes() == arr.newbyteorder(">").tobytes()
Out[21]: True

In [22]: arr.astype(dt_big).tobytes() == arr.astype(dt_little).tobytes()
Out[22]: False
```

And we can use this to bypass the need to actually test on little-endian, since calling `original.astype(dt_little)` will change the array into what would have been produced on that hardware (?).

And the current code fails in both cases:

```
In [25]: serialize_numpy(arr.astype(dt_big))[0]
Out[25]: b'\x00\x00\x00\x00\x00\x00\x00\x00?\xf0\x00\x00\x00\x00\x00\x00@\x00\x00\x00\x00\x00\x00\x00'

In [26]: serialize_numpy(arr.astype(dt_little))[0]
Out[26]: b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf0?\x00\x00\x00\x00\x00\x00\x00@'
```

Different if you look closely.
```

In [35]: deserialize_numpy(arr.astype(dt_big).tobytes(), shape=(3, 1))
Out[35]:
array([[0.],
       [1.],
       [2.]])

In [36]: deserialize_numpy(arr.astype(dt_little).tobytes(), shape=(3, 1))
Out[36]:
array([[0.00000e+000],
       [3.03865e-319],
       [3.16202e-322]])
```

the last line here in principle should be exhibiting what would happen if the current code was called on a PowerPC (?) machine.

**Comment by mattwthompson** [comment]
Also FTR the units refactor hit some serialization code, but not these specific bits which are only hit after units have been stripped or before units are attached. A fix here should be mergable into the refactor branch.

**Comment by lilyminium** [comment]
> So calling .newbyteorder on an array doesn't produce the effect we wanted (make sure the bytes are in the expected order). But calling it on a type does (?)

That sounds like how I understand it -- `newbyteorder` doesn't do anything about the data in  memory but just changes the reading frame, whereas `astype` will actually create a new array in memory if the existing array is not already of the requested type.

> And we can use this to bypass the need to actually test on little-endian, since calling original.astype(dt_little) will change the array into what would have been produced on that hardware (?).

I reckon so -- from my limited worldview I've only ever used little-endian architectures, but by creating big-endian objects we can use that to test. This function is a little easier to test than the one in interchange because the native architecture (for me) is little, and the serialization functions convert to/from big. It looks like you're natively big-endian? I get the opposite behaviour 😆 (which actually shows exactly why this function should exist!)

```python
>>> deserialize_numpy(arr.astype(dt_big).tobytes(), shape=(3, 1))
array([[0.00000e+000],
       [3.03865e-319],
       [3.16202e-322]])
deserialize_numpy(arr.astype(dt_little).tobytes(), shape=(3, 1))
array([[0.],
       [1.],
       [2.]])
```

**Comment by lilyminium** [comment]
No worries! I am in NYC at the moment, though, so any feedback might be delayed.

**Comment by mattwthompson** [comment]
Fixed/addressed in #1155, expected to be a part of v0.10.2
